python/runfiles/runfiles.py

```diff
@@ -15,12 +15,126 @@
+from typing import Dict, Iterator, Optional, Tuple, Union
+
+
+class _RepositoryMapping(collections.abc.Mapping):
+    """Repository mapping for resolving apparent repository names to canonical ones.
+    
+    Handles both exact mappings and prefix-based mappings introduced by the 
+    --incompatible_compact_repo_mapping_manifest flag.
+    """
+
+    def __init__(
+        self, 
+        exact_mappings: Dict[Tuple[str, str], str], 
+        prefixed_mappings: Dict[Tuple[str, str], str]
+    ) -> None:
+        """Initialize repository mapping with exact and prefixed mappings.
+        
+        Args:
+            exact_mappings: Dict mapping (source_canonical, target_apparent) -> target_canonical
+            prefixed_mappings: Dict mapping (source_prefix, target_apparent) -> target_canonical
+        """
+        self._exact_mappings = exact_mappings
+        self._prefixed_mappings = prefixed_mappings
+        
+        # Group prefixed mappings by target_apparent for faster lookups
+        self._grouped_prefixed_mappings = defaultdict(list)
+        for (prefix_source, target_app), target_canonical in self._prefixed_mappings.items():
+            self._grouped_prefixed_mappings[target_app].append((prefix_source, target_canonical))
+
+    @staticmethod
+    def create_from_file(repo_mapping_path: Optional[str]) -> "_RepositoryMapping":
+        """Create RepositoryMapping from a repository mapping manifest file.
+        
+        Args:
+            repo_mapping_path: Path to the repository mapping file, or None if not available
+            
+        Returns:
+            RepositoryMapping instance with parsed mappings
+        """
+        # If the repository mapping file can't be found, that is not an error: We
+        # might be running without Bzlmod enabled or there may not be any runfiles.
+        # In this case, just apply empty repo mappings.
+        if not repo_mapping_path:
+            return _RepositoryMapping({}, {})
+        
+        try:
+            with open(repo_mapping_path, "r", encoding="utf-8", newline="\n") as f:
+                content = f.read()
+        except FileNotFoundError:
+            return _RepositoryMapping({}, {})
+
+        exact_mappings = {}
+        prefixed_mappings = {}
+        for line in content.splitlines():
+            source_canonical, target_apparent, target_canonical = line.split(",")
+            if source_canonical.endswith("*"):
+                # This is a prefixed mapping - remove the '*' for prefix matching
+                prefix = source_canonical[:-1]
+                prefixed_mappings[(prefix, target_apparent)] = target_canonical
+            else:
+                # This is an exact mapping
+                exact_mappings[(source_canonical, target_apparent)] = target_canonical
+
+        return _RepositoryMapping(exact_mappings, prefixed_mappings)
+
+    # Mapping protocol implementation
+    def __getitem__(self, key: Tuple[str, str]) -> str:
```

Comment:
Please rename this to e.g. `get_canonical` or similar, just not `__getitem__`. This is an internal custom class, so there's not need to add the abstraction or complications of making it look dict-like.